### PR TITLE
Implement snake game

### LINKED ISSUE #39: Implement Snake game
# :white_check_mark: What's needed
You must implement the Snake Game.
It consists of:
- [x] Init the map (by creating Entity Classes)
- [x] Init the Snake player (by creating a Player Class)
- [x] Init the spawn of apples 
- [x] Implement the key events
- [x] Fill the little functions (`start`, `stop`, `getScore`, etc...)
- [x] Complete the `simulate` function

## :page_facing_up: Basic Rules
- [x] The game area is a finite amount of cells. The edges of the area either cannot be passed through or
are cyclical.
- [x] The snake starts with a size of 4 cells in the middle of the area.
- [x] The snake moves forward automatically at a constant speed. Each section of its tail follows the exact
same path as the head.
- [x] The snake can turn right or left when the corresponding key is pressed.
- [x] The goal of the game is to feed the snake so that it can grow. The game area MUST NEVER have less
than one element of food.
- [x] A food element fills a single cell.
- [x] When the head of the snake goes over a cell with food, the food disappears and a one-cell-long
section is added at the tail of the snake. The new section appears in the first free tile next to the last
cell of the tail. If there is no free cell, the game is over. If a new section is added, a new food element
appears.
- [x] When the head of the snake runs into the border of the screen (if the map isn’t cyclical) or a part of
its body, the game is over

## :file_folder: Files required
- [x] `src/Games/SNAKE/Snake.cpp`
- [x] `src/Games/SNAKE/Snake.hpp`

### PR DESCRIPTION
**Pull request by Thomaltarix**
I finished the Snake Game completely.
To see details, check this issue:
- #39

### CODE REVIEW

File: src/Core/Core.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 */
 
 #include "Core.hpp"
+#include "IGame.hpp"
```
Comment (mathieurobert1) [inline src/Core/Core.cpp]: IGame is already in the Core.hpp

File: src/Games/AGame.hpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+                    const char* what() const noexcept override
+                    {
+                        return _msg.c_str();
+                    };
+
+                private:
+                    std::string _msg;
+            };
+            private:
```
Comment (mathieurobert1) [inline src/Games/AGame.hpp]: Protected instead of Private in an Abstract class ?

File: src/Games/APlayer.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/*
+** EPITECH PROJECT, 2024
+** Arrrrrcade
+** File description:
+** APlayer
+*/
+
+#include "APlayer.hpp"
+
+int Arcade::APlayer::die()
```
Comment (mathieurobert1) [inline src/Games/APlayer.cpp]: why int function if the return is always -1 ?

File: src/Games/APlayer.hpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+                    const char* what() const noexcept override
+                    {
+                        return _msg.c_str();
+                    };
+
+                private:
+                    std::string _msg;
+            };
+        private:
```
Comment (mathieurobert1) [inline src/Games/APlayer.hpp]: it's not protected for an Abstract ?

File: src/Games/SNAKE/Snake.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    return _texts;
+}
+
+std::vector<std::shared_ptr<Arcade::ISound>> Arcade::SnakeGame::getSounds()
+{
+    return _sounds;
+}
+
+
+void Arcade::SnakeGame::initMap(int width, int height)
```
Comment (mathieurobert1) [inline src/Games/SNAKE/Snake.cpp:113]: The map initialization is not good, look at the Centipede map creation to create the map

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
Some changes to makes but a really good code

File: src/Core/Core.cpp (hunk 1 above)
Comment (Thomaltarix) [reply src/Core/Core.cpp]: Done

File: src/Games/AGame.hpp (hunk 2 above)
Comment (Thomaltarix) [reply src/Games/AGame.hpp]: Yes, you're right !

File: src/Games/APlayer.hpp (hunk 4 above)
Comment (Thomaltarix) [reply src/Games/APlayer.hpp]: Same as above, you're right here too

File: src/Games/APlayer.cpp (hunk 3 above)
Comment (Thomaltarix) [reply src/Games/APlayer.cpp]: Good request !

File: src/Games/SNAKE/Snake.cpp (hunk 5 above)
Comment (Thomaltarix) [reply src/Games/SNAKE/Snake.cpp:113]: I think we are just reversing the x and y for our map, so it causes nothing

**Review by mathieurobert1 — APPROVED** [review]
Seems perfect :+1:

**Pull request merged**